pkg/executor/sortexec/sort.go

```diff
@@ -336,11 +474,112 @@ func (e *SortExec) fetchRowChunks(ctx context.Context) error {
 		return err
 	}
 
-	e.sortPartitions = append(e.sortPartitions, e.curPartition)
+	e.Unparallel.sortPartitions = append(e.Unparallel.sortPartitions, e.curPartition)
 	e.curPartition = nil
 	return nil
 }
 
+func (e *SortExec) fetchChunksParallel(ctx context.Context) error {
+	workerNum := len(e.Parallel.workers)
+
+	// Wait for the finish of all workers
+	workersWaiter := sync.WaitGroup{}
+
+	// Add before the start of goroutine to avoid that the counter is minus to negative.
+	workersWaiter.Add(workerNum + 1)
+
+	// Fetch chunks from child and put chunks into MPMCQueue
+	go e.fetchChunksFromChild(ctx, &workersWaiter)
+
+	// Create and run workers
+	for i := range e.Parallel.workers {
+		e.Parallel.workers[i] = newParallelSortWorker(i, e.lessRow, e.Parallel.globalSortedRowsQueue, &workersWaiter, &e.Parallel.result, e.Parallel.chunkChannel, e.tryToCloseChunkChannel, e.checkErrorForParallel, e.processErrorForParallel, e.memTracker)
+		go e.Parallel.workers[i].run()
+	}
+
+	// Wait for the finish of all workers
+	workersWaiter.Wait()
+	err := e.checkErrorForParallel()
+	if err != nil {
+		return err
+	}
+	e.fetchResultFromQueue()
+	return nil
+}
+
+// Fetch chunks from child and put chunks into MPMCQueue
+func (e *SortExec) fetchChunksFromChild(ctx context.Context, waitGroup *sync.WaitGroup) {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(e.processErrorForParallel, r)
+		}
+		e.tryToCloseChunkChannel()
+		waitGroup.Done()
+	}()
+
+	for {
+		chk := exec.TryNewCacheChunk(e.Children(0))
+		err := exec.Next(ctx, e.Children(0), chk)
+		if err != nil {
+			e.processErrorForParallel(err)
+			return
+		}
+
+		rowCount := chk.NumRows()
+		if rowCount == 0 {
+			break
+		}
+
+		chkWithMemoryUsage := &chunkWithMemoryUsage{
+			Chk:         chk,
+			MemoryUsage: chk.MemoryUsage() + chunk.RowSize*int64(rowCount),
```

Comment:
why `chunk.RowSize*int64(rowCount)` is needed, doesn't `chk.MemoryUsage()` contains it?

Reply:
> why `chunk.RowSize*int64(rowCount)` is needed, doesn't `chk.MemoryUsage()` contains it?

Because we will extract rows from chunks and save them in a field with `[]chunk.Row` type, memory consumed by `chunk.Row` is also needed to be tracked.

Reply:
`getRows` is called in `sortChunkAndGetSortedRows`, so `memory_tracker.consume(chunk.RowSize*int64(rowCount))` should be called in `sortChunkAndGetSortedRows` instead of here?

Reply:
> `getRows` is called in `sortChunkAndGetSortedRows`, so `memory_tracker.consume(chunk.RowSize*int64(rowCount))` should be called in `sortChunkAndGetSortedRows` instead of here?

I think it's better to call at here, as memory consumed by `chunk.Row` must be allocated in the future and we can call `Consume()` only once.

Reply:
Tracker accuracy might be more important?
The overhead of tracker.consume seems small.
Does chunk's memory_used need to be subtracted in `sortChunkAndGetSortedRows`?